docs/design/docker-deploy.md

```diff
@@ -0,0 +1,651 @@
+1. User is free to use the Angular CLI without Docker. By default, do not generate Docker files upon creation of a new project (`ng new`, `ng init`).
+1. Don't recreate the wheel. Implement the common Docker use cases that make it convenient for Angular applications.
+1. Don't inhibit users from using the standalone Docker CLI tools for other use cases.
+1. Assumes 1:1 Dockerfile with Angular project. Other services should be outside (or sibling to) the ng app project folder.
+1. Generated starter Dockerfile will use the current official Node.js LTS base image.
+1. Generated Dockerfile will use `npm run docker-start` as its default CMD. Requires new `package.json` script.
+1. User is free to modify and customize all generated files directly without involvement by the Angular CLI.
+1. Image builds will support all Docker build options.
+1. Container deploys will support all Docker run options.
+1. Deploying to a Docker Machine can be local or remote.
+1. Deploys can be made to different environments (dev, stage, prod) on the same or different Docker Machines.
+1. Image pushes can be made to Docker Hub, AWS ECR, and other public/private registries.
+1. Adhere to [Docker security best practices](https://docs.docker.com/engine/security/).
+1. Use sensible defaults to make it easy for users to get started.
+1. Utilize ngConfig project config file to save information for successive `docker:*` command runs.
+1. Support `--dry-run` and `--verbose` flags.
+
+
+## Proposed CLI API
+
+### Overview
+
+Initialize Docker files and configuration for a default build and run environment:
+```
+$ ng docker:init
+```
+
+Initialize a docker-compose file for deploying to a separate environment:
+```
+$ ng docker:init {environment}
+```
+
+Build and push a Docker image of the Angular app to the registry:
+```
+$ ng docker:push
+```
+
+Deploy and run the Angular app on a Docker Machine:
+```
+$ ng docker:deploy [environment]
+```
+
+Display logs of a running deployment:
+```
+$ ng docker:logs [environment] [--services service]
+```
+
+Attach to a running service container and run a command:
+```
+$ ng docker:exec [environment] [--service service] [--cmd bash]
+```
+
+### Command - Initialize Default Build and Run Environment
+
+The command `docker:init` prompts the user for configuration options and generates starter `Dockerfile`, `.dockerignore`, and `docker-compose.yml` files for building/pushing images to a registry (via `docker:push`) and running (via `docker:deploy`).
+
+Most users will start with one local 'default' Docker Machine (Mac/Win), or a local native Docker environment on Linux, where they will perform builds and run containers for development testing. This command prepares the Angular project for working with that default environment.
+
+Example - Init default build environment:
+```bash
+$ ng docker:init
+> Which Docker Machine for the default "build and run" environment?
+  [*] default / local
```

Comment:
Is there any case where this would be different?

I'm kinda adverse to prompts. If there's a configuration it should be a flag with a sensible default.